src/VisualStudio/Core/Def/ProjectSystem/VisualStudioProjectFactory.cs

```diff
@@ -79,6 +79,10 @@ public Task<VisualStudioProject> CreateAndAddToWorkspaceAsync(string projectSyst
             // From this point on, we start mutating the solution.  So make us non cancellable.
             cancellationToken = CancellationToken.None;
 
+            // Move the rest of this work off from the UI thread; although it should be fast, sometimes an accidental MEF part creation
+            // can slow things down
+            await TaskScheduler.Default;
```

Comment:
So I don't like this -- for the CPS case, this is a benefit. For the legacy project system case, this method is being called under a JTF.Run() on the UI thread, so all this is going to do is slow things down. How do we get a best-of-both-worlds here?

Reply:
> How do we get a best-of-both-worlds here?

Optimizing for the desired stable state is generally the best approach. Once the caller in legacy project system is made asynchronous or eliminated, we automatically reach the goal.